File: 205-IsomorphicStrings.java
public class Solution {
    public boolean isIsomorphic(String s, String t) {
        if (s == null || t == null) return false;
        if (s.length() != t.length())   return false;
        if (s == null || t == null) return false;
        HashMap<Character, Character> hm = new HashMap<>();
        
        for (int i = 0; i < s.length(); i++){
            if (hm.containsKey(s.charAt(i))){
                if (hm.get(s.charAt(i)) == t.charAt(i)){
                    continue;
                }else{
                    return false;
                }
            }else{
                if (!hm.containsValue(t.charAt(i))){
                    hm.put(s.charAt(i), t.charAt(i));
                }else{
                    return false;
                }
            }
        }
        return true;
    }
}

File: cpp/205-IsomorphicStrings.cpp
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        unordered_map<char, char> st;
        unordered_map<char, char> ts;
        if (s.size() != t.size())
            return false;
        for (int i = 0; i < s.size(); i++) {
            if (st.find(s[i]) == st.end() && ts.find(t[i]) == ts.end()) {
                st[s[i]] = t[i];
                ts[t[i]] = s[i];
            } else if (st[s[i]] != t[i] || ts[t[i]] != s[i]) {
                return false;
            }
        }
        return true;
    }
};
